### ISSUE TITLE
Previously made subscriptions with same observer are unsubscribed when subsequent subscription completes

### ISSUE THREAD

**Issue by oscarlorentzon** [ReactiveX/rxjs#2237]
**RxJS version:**

`5.0.2`

**Code to reproduce:**

```ts
let source = Observable
    .timer(1000, 1000)
    .take(2);

let observer = {
    next: () => { }
};

source
    .do(v => console.log("a", v))
    .subscribe(observer);

source
    .do(v => console.log("b", v))
    .subscribe({
        next: () => { }
    });

Observable.empty()
    .subscribe(observer);
```

**Expected behavior:**

Expected console output:
```js
a 0
b 0
a 1
b 1
```

**Actual behavior:**

Actual console output:
```js
b 0
b 1
```

**Additional information:**

The root cause for this behavior is the following logic:

https://github.com/ReactiveX/rxjs/blob/6922b16f4a297a4acbe8ad2b086c947b7c0d8315/src/Subscriber.ts#L174-L177

The observer instance in the example above will get the unsubscribe function of the `Subscriber` attached to it in the first source `subscribe` call. In the final `subscribe` call to the empty observable the observer instance already has the attached unsubscribe function which will be added to the `Subscription`. When the empty observable subscription completes, which happens immediately, a chain reaction will occur. The empty observable will unsubscribe on complete and it will subsequently call the previously added unsubscribe function which is related to the first source subscription. That subsription will also be unsubscribed and closed.

Another example of the problem is when `subscribe` is called with no arguments:

```ts
let source = Observable
    .timer(1000, 1000)
    .take(2);

source
    .do(v => console.log("a", v))
    .subscribe();

source
    .do(v => console.log("b", v))
    .subscribe(() => { });

Observable.empty()
    .subscribe();
```

Expected console output:
```js
a 0
b 0
a 1
b 1
```

Actual console output:
```js
b 0
b 1
```

This was introduced in `5.0.0-rc.1` and is caused by the fact that the `Observer.empty` instance is used for all calls to `toSubcriber` with no arguments:

https://github.com/ReactiveX/rxjs/blob/6922b16f4a297a4acbe8ad2b086c947b7c0d8315/src/util/toSubscriber.ts#L21

This logic also means that a larger chain reaction can occur when multiple calls to subscribe with no arguments on the same or different source observables are made. If the last subscription completes, all previous subscriptions that were made with no arguments will be unsubscribed and closed.

The fact that the library behaves differently when using a next function instance (see below) as opposed to an observer object instance indicates that the behavior described above must be a bug.

```ts
let source = Observable
    .timer(1000, 1000)
    .take(2);

let next = () => { };

source
    .do(v => console.log("a", v))
    .subscribe(next);

source
    .do(v => console.log("b", v))
    .subscribe(() => { });

Observable.empty()
    .subscribe(next);
```

Actual console output:
```js
a 0
b 0
a 1
b 1
```

**Comment by trxcllnt** [comment]
Yes, I believe SafeSubscriber needs to `Object.create(observer)` the incoming Observer, instead of mutating it. This has never been safe, and was only meant to be temporary.

**Comment by mattflix** [comment]
I just spent a long time coming up with a repro case for some very weird stuff my app was doing (at which point I finally figured out the underlying unsubscribe behavior described in the above issue) then only to discover the issue is only already known.

Here is the repro:  http://jsbin.com/bitogaz/2/edit?html,js,console

In my case, the "same observers" are actually no observers at all (i.e., I was calling subscribe() with no arguments).  Not sure if the other repros demonstrate this particular case.

**Comment by trxcllnt** [comment]
@mattflix will be fixed when #2238 is merged

**Comment by oscarlorentzon** [comment]
@mattflix The second example above demonstrates the subscribe call with no arguments case.

**Comment by kwonoj** [comment]
Closed via #2238.
